src/file-upload/file-list.tsx

```diff
@@ -0,0 +1,237 @@
+    const [itemsRenderMode, setItemsRenderMode] = useState<
+        FileItemRenderProps[]
+    >([]);
+
+    const { width: wrapperWidth, ref: wrapperRef } = useResizeDetector();
+
+    // =========================================================================
+    // EFFECTS
+    // =========================================================================
+    useEffect(() => {
+        setItemsRenderMode(getItemsRenderMode(fileItems));
+    }, [fileItems]);
+
+    useEffect(() => {
+        setArrangedItems(getArrangedItemsToRender(itemsRenderMode));
+    }, [itemsRenderMode]);
+
+    // =========================================================================
+    // EVENT HANDLERS
+    // =========================================================================
+    const handleDescriptionUpdate =
+        (item: FileItemProps) => (description: string) => {
+            const updatedItem = { ...item, description };
+            onItemUpdate(updatedItem);
+        };
+
+    const handleCancel = (item: FileItemProps) => () => {
+        setItemsRenderMode((prevState) => {
+            return prevState.map((renderItem) => {
+                if (isEqual(item, renderItem.fileItem)) {
+                    return {
+                        ...renderItem,
+                        renderMode: item.description
+                            ? "list"
+                            : "cancelled-edit",
+                    };
+                }
+
+                return renderItem;
+            });
+        });
+    };
+
+    const handleInitiateEdit = (item: FileItemProps) => () => {
+        setItemsRenderMode((prevState) => {
+            return prevState.map((renderItem) => {
+                if (isEqual(item, renderItem.fileItem)) {
+                    return {
+                        ...renderItem,
+                        renderMode: "edit",
+                    };
+                }
+
+                return renderItem;
+            });
+        });
+    };
+
+    const handleDelete = (item: FileItemProps) => () => {
+        onItemDelete(item);
+    };
+
+    // =========================================================================
+    // HELPER FUNCTIONS
+    // =========================================================================
+    const checkEditable = (item: FileItemProps) => {
+        return (
+            editableFileItems &&
+            FileUploadHelper.isSupportedImageType(item.type)
+        );
+    };
+
+    const shouldRenderEditMode = (item: FileItemProps) => {
+        return checkEditable(item) && item.description === undefined;
+    };
+
+    const getItemsRenderMode = (
+        fileItems: FileItemProps[]
+    ): FileItemRenderProps[] => {
+        if (!fileItems || fileItems.length === 0) return [];
+
+        return fileItems.map((item) => {
+            const existingIndex = findIndex(itemsRenderMode, (renderedItem) => {
+                return isEqual(renderedItem.fileItem, item);
```

Comment:
should compare the file item id here, otherwise a description change would also count as a new item. for existing rendered items I think it could just retain the current `renderMode`

Reply:
Actually i am looking to rewrite this. Let me refactor